src/lib/profiles/data-management/Current/CommandSender.cpp

```diff
@@ -0,0 +1,537 @@
+    }
+
+    dataSinkVersion = mDataSink->GetVersion();
+
+    if (mPostCommandVersion == 0)
+    {
+        return false;
+    }
+
+    if (mPreCommandVersion == 0)
+    {
+        mPreCommandVersion = mPostCommandVersion - kCommandSideEffectWindowSize;
+    }
+
+    {
+        bool newerThanPost = (dataSinkVersion >= mPostCommandVersion);
+        bool olderThanPre = (dataSinkVersion < mPreCommandVersion);
+
+        if (mPostCommandVersion > mPreCommandVersion)
+        {
+            if (newerThanPost || olderThanPre)
+            {
+                return true;
+            }
+            else
+            {
+                return false;
+            }
+        }
+        else
+        {
+            if (newerThanPost && olderThanPre)
+            {
+                return true;
+            }
+            else
+            {
+                return false;
+            }
+        }
+    }
+}
+
+/**
+ * Initializes the trait path for the command (within the arguments object) given a TraitDataSink + TraitCatalog housing that sink and the commadn type.
+ *
+ * @param[in] aCatalog          Pointer to the TraitCatalog housing the data sink
+ * @param[in] aSink             Pointer to the trait data sink
+ * @param[in] aCommandType      Type of command
+ *
+ * @retval WEAVE_ERROR          WEAVE_NO_ERROR if success, error otherwise.
+ */
+WEAVE_ERROR CommandSender::SendParams::PopulateTraitPath(TraitCatalogBase<TraitDataSink> *aCatalog, TraitDataSink *aSink, uint32_t aCommandType)
+{
+    WEAVE_ERROR err = WEAVE_NO_ERROR;
+    TraitDataHandle handle;
+
+    err = aCatalog->Locate(aSink, handle);
+    SuccessOrExit(err);
+
+    err = aCatalog->GetInstanceId(handle, InstanceId);
+    SuccessOrExit(err);
+
+    err = aCatalog->GetResourceId(handle, ResourceId);
+    SuccessOrExit(err);
+
+    ProfileId = aSink->GetSchemaEngine()->GetProfileId();
+    VersionRange.mMaxVersion = aSink->GetSchemaEngine()->GetMaxVersion();
+    VersionRange.mMinVersion = aSink->GetSchemaEngine()->GetMinVersion();
+    CommandType = aCommandType;
+
+exit:
+    WeaveLogFunctError(err);
+    return err;
+}
+
+/**
+ * Initializes some of the appropriate fields in the CommandSender object.
+ *
+ * @param[in] aBinding          Binding that should be used with this command for all future sends.
+ * @param[in] aEventCallback    Callback for the application register interest in notable events. Can be NULL
+ * @param[in] aAppState         App context to be passed into the event callback
+ *
+ * @retval WEAVE_ERROR          WEAVE_ERROR_DEFAULT_EVENT_HANDLER_NOT_CALLED if the default handler isn't setup.
+ *                              Otherwise, WEAVE_NO_ERROR
+ */
+WEAVE_ERROR CommandSender::Init(Binding *aBinding, const EventCallback aEventCallback, void * const aAppState)
+{
+    WEAVE_ERROR err = WEAVE_NO_ERROR;
+
+    VerifyOrExit(aBinding != NULL, err = WEAVE_ERROR_INVALID_ARGUMENT);
+
+    mBinding = aBinding;
+    mBinding->AddRef();
+
+    if (aEventCallback)
```

Comment:
from previous PR (@jaylogue):

It should be an error for event callback to be NULL.

Reply:
@mrjerryjohns 

What was intended here? Is the DefaultEventHandler supposed to be used if EventCallback is NULL? Or is that a separate function?